File: py/795.number-of-subarrays-with-bounded-maximum.py
#
# @lc app=leetcode id=795 lang=python3
#
# [795] Number of Subarrays with Bounded Maximum
#

# @lc code=start
class Solution:
    def numSubarrayBoundedMax(self, nums: List[int], left: int, right: int) -> int:
        count = 0
        confirmed = False
        prev_confirmed_distance = 0
        res = 0
        for n in nums:
            if n <= right:
                count += 1
                if n >= left:
                    res += count
                    confirmed = True
                    prev_confirmed_distance = 0
                elif confirmed:
                    prev_confirmed_distance += 1
                    res += count-prev_confirmed_distance
            else:
                count = 0
                confirmed = False
        return res
# @lc code=end

File: 795.number-of-subarrays-with-bounded-maximum.cpp
/*
 * @lc app=leetcode id=795 lang=cpp
 *
 * [795] Number of Subarrays with Bounded Maximum
 */
#include <vector>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
public:
    int numSubarrayBoundedMax(vector<int> &nums, int left, int right)
    {
        int count = 0;
        bool pass = false;
        int prev_pass_distance = 0;
        int res = 0;
        for (auto &&n : nums)
        {
            if (n <= right)
            {
                ++count;
                if (n >= left)
                {
                    res += count;
                    pass = true;
                    prev_pass_distance = 0;
                }
                else if (pass)
                {
                    ++prev_pass_distance;
                    res += count - prev_pass_distance;
                }
            }
            else
            {
                count = 0;
                pass = false;
            }
        }
        return res;
    }
};
// @lc code=end